planner/core/exhaust_physical_plans.go

```diff
@@ -387,6 +387,101 @@ func (p *LogicalJoin) constructIndexJoin(
 	return []PhysicalPlan{join}
 }
 
+func findIndexScanAndTableScan(p PhysicalPlan) (is *PhysicalIndexScan, ts *PhysicalTableScan) {
+	var tblChildren []PhysicalPlan
+	var idxChildren []PhysicalPlan
+	switch x := p.(type) {
+	case *PhysicalTableReader:
+		tblChildren = x.TablePlans
+	case *PhysicalIndexReader:
+		idxChildren = x.IndexPlans
+	case *PhysicalIndexLookUpReader:
+		tblChildren = x.TablePlans
+		idxChildren = x.IndexPlans
+	default:
+		tblChildren = x.Children()
+	}
+	for _, child := range tblChildren {
+		tmpIS, tmpTS := findIndexScanAndTableScan(child)
+		if tmpIS != nil {
+			is = tmpIS
+		}
+		if tmpTS != nil {
+			ts = tmpTS
+		}
+	}
+	for _, child := range idxChildren {
+		tmpIS, tmpTS := findIndexScanAndTableScan(child)
+		if tmpIS != nil {
+			is = tmpIS
+		}
+		if tmpTS != nil {
+			ts = tmpTS
+		}
+	}
+	if tmpIS, ok := p.(*PhysicalIndexScan); ok {
+		return tmpIS, ts
+	}
+	if tmpTS, ok := p.(*PhysicalTableScan); ok {
+		return is, tmpTS
+	}
+	return is, ts
+}
+
+func (p *LogicalJoin) constructIndexMergeJoin(
+	prop *property.PhysicalProperty,
+	outerIdx int,
+	innerPlan PhysicalPlan,
+	ranges []*ranger.Range,
+	keyOff2IdxOff []int,
+	lens []int,
+	compareFilters *ColWithCmpFuncManager,
+) []PhysicalPlan {
+	indexJoins := p.constructIndexJoin(prop, outerIdx, innerPlan, ranges, keyOff2IdxOff, lens, compareFilters)
+	indexMergeJoins := make([]PhysicalPlan, 0, len(indexJoins))
+	is, ts := findIndexScanAndTableScan(innerPlan)
+	for _, plan := range indexJoins {
+		join := plan.(*PhysicalIndexJoin)
+		// needOuterSort means whether the outer join keys are the prefix of the prop items.
+		needOuterSort := len(prop.Items) > 0 && len(join.OuterJoinKeys) <= len(prop.Items)
+		compareFuncs := make([]expression.CompareFunc, 0, len(join.OuterJoinKeys))
+		outerCompareFuncs := make([]expression.CompareFunc, 0, len(join.OuterJoinKeys))
+		for i := range join.OuterJoinKeys {
+			if needOuterSort && !prop.Items[i].Col.Equal(nil, join.OuterJoinKeys[i]) {
+				needOuterSort = false
+			}
+			compareFuncs = append(compareFuncs, expression.GetCmpFunction(join.OuterJoinKeys[i], join.InnerJoinKeys[i]))
+			outerCompareFuncs = append(outerCompareFuncs, expression.GetCmpFunction(join.OuterJoinKeys[i], join.OuterJoinKeys[i]))
+		}
+		// canKeepOuterOrder means whether the prop items are the prefix of the outer join keys.
+		canKeepOuterOrder := len(prop.Items) <= len(join.OuterJoinKeys)
+		for i := 0; canKeepOuterOrder && i < len(prop.Items); i++ {
+			if !prop.Items[i].Col.Equal(nil, join.OuterJoinKeys[i]) {
+				canKeepOuterOrder = false
+			}
+		}
+		// If the prop items are NOT the prefix of the outer join keys,
```

Comment:
You can use comment like `since index merge join requires..., so we need one of them be true`, now this comment adds no more infomation than the code.